src/sage/symbolic/relation.py

```diff
@@ -1026,108 +1117,92 @@ def solve(f, *args, **kwds):
         sage: solve([1], x)
         Traceback (most recent call last):
         ...
-        TypeError: The first argument to solve() should be a symbolic expression
-        or a list of symbolic expressions.
-    """
-    from sage.structure.element import Expression
-    explicit_solutions = kwds.get('explicit_solutions', None)
-    multiplicities = kwds.get('multiplicities', None)
-    to_poly_solve = kwds.get('to_poly_solve', None)
-    solution_dict = kwds.get('solution_dict', False)
-    algorithm = kwds.get('algorithm', None)
-    domain = kwds.get('domain', None)
-
-    if len(args) > 1:
-        x = args
-    else:
-        x = args[0]
-    if isinstance(x, (list, tuple)):
-        for i in x:
-            if not isinstance(i, Expression):
-                raise TypeError("%s is not a valid variable." % repr(i))
-    elif x is None:
-        vars = f.variables()
-        if len(vars) == 0:
-            if multiplicities:
-                return [], []
-            else:
-                return []
-        x = vars[0]
-    elif not isinstance(x, Expression):
-        raise TypeError("%s is not a valid variable." % repr(x))
-
-    if isinstance(f, (list, tuple)) and len(f) == 1:
-        # f is a list with a single element
-        if isinstance(f[0], Expression):
-            f = f[0]
-        else:
-            raise TypeError("The first argument to solve() should be a "
-                            "symbolic expression or a list of symbolic "
-                            "expressions.")
-
-    if isinstance(f, Expression):  # f is a single expression
-        return _solve_expression(f, x, explicit_solutions, multiplicities, to_poly_solve, solution_dict, algorithm, domain)
+        TypeError: must be a symbolic expression or a list of symbolic expressions
 
-    if not isinstance(f, (list, tuple)):
-        raise TypeError("The first argument must be a symbolic expression or a list of symbolic expressions.")
+    Automatic computation of the variables to solve for::
 
-    # f is a list of such expressions or equations
+        sage: var("x y")
+        (x, y)
+        sage: solve(x==1)
+        [x == 1]
+        sage: solve([x == 1, y == 2])
+        [[y == 2, x == 1]]
+        sage: solve([x == 1])
+        [x == 1]
+        sage: solve(x == 1)
+        [x == 1]
+        sage: solve([x - x == 1])
+        []
+        sage: solve(x - x == 1)
+        []
+        sage: solve([x - x == 0])
+        []
+        sage: solve(x - x == 0)
+        []
```

Comment:
When no variables are provided, by right the return value should be `[[]]` if at least a solution exists. Is this a important point?